command-not-found.sh

```diff
@@ -4,36 +4,97 @@
 # this will be called when a command is entered
 # but not found in the user’s path + environment
 command_not_found_handle () {
-    toplevel=nixos # TODO: detect this somehow
-    cmd=$1 # TODO: differentiate between paths and commands
+
+    # TODO: use "command not found" gettext translations
+
+    # taken from http://www.linuxjournal.com/content/bash-command-not-found
+    # - do not run when inside Midnight Commander or within a Pipe
+    if [ -n "$MC_SID" ] || ! [ -t 1 ]; then
+        >&2 echo "$1: command not found"
+        return 127
+    fi
+
+    toplevel=nixpkgs # nixpkgs should always be available even in NixOS
+    cmd=$1
     attrs=$(@out@/bin/nix-locate --minimal --no-group --type x --top-level --whole-name --at-root "/bin/$cmd")
     len=$(echo -n "$attrs" | grep -c "^")
 
     case $len in
         0)
             >&2 echo "$cmd: command not found"
             ;;
-        1) # TODO: add option to autorun with 1 match
-            >&2 echo "The program '$cmd' is currently not installed. You can install it"
-            >&2 echo "by typing:"
-            >&2 echo "  nix-env -iA $toplevel.$attrs"
+        1)
+            # if only 1 package provides this, then we can invoke it
+            # without asking the users if they have opted in with one
+            # of 2 environment variables
+
+            # they are based on the ones found in
+            # command-not-found.sh:
+
+            #   NIX_AUTO_INSTALL : install the missing command into the
+            #                      user’s environment
+            #   NIX_AUTO_RUN     : run the command transparently inside of
+            #                      nix shell
+
+            # these will not return 127 if they worked correctly
+
+            if ! [ -z "$NIX_AUTO_INSTALL" ]; then
+                >&2 cat <<EOF
+The program '$cmd' is currently not installed. It is provided by
+the package '$toplevel.$attrs', which I will now install for you.
+EOF
+                nix-env -iA $toplevel.$attrs
+                if [ "$?" -eq 0 ]; then
+                    $@ # TODO: handle pipes correctly if AUTO_RUN/INSTALL is possible
```

Comment:
Is this how `command-not-found` works in NixOS? From looking at the source at https://github.com/NixOS/nixpkgs/blob/a35ec5dda688f6eedbb281671ee11261c769610e/nixos/modules/programs/command-not-found/command-not-found.pl#L34, it seems it does not run the program, only installs it and exits with `126`.